### PR TITLE
Spotless + Ktlint + Compose rules

### PR DESCRIPTION
**Pull request by Goooler**

### CODE REVIEW

**Comment by Goooler** [comment]
Introduce `org.jetbrains.kotlinx:kotlinx-collections-immutable` to fit https://mrmans0n.github.io/compose-rules/rules/#use-immutable-annotation-whenever-possible

```
OLD: before.apk (signature: V2)
NEW: after.apk (signature: V2)

          │           compressed            │          uncompressed
          ├──────────┬──────────┬───────────┼──────────┬──────────┬───────────
 APK      │ old      │ new      │ diff      │ old      │ new      │ diff
──────────┼──────────┼──────────┼───────────┼──────────┼──────────┼───────────
      dex │  2.2 MiB │  2.2 MiB │ +13.7 KiB │  2.2 MiB │  2.2 MiB │ +13.7 KiB
     arsc │ 21.1 KiB │ 21.1 KiB │       0 B │   21 KiB │   21 KiB │       0 B
 manifest │  2.9 KiB │  2.9 KiB │       0 B │ 12.9 KiB │ 12.9 KiB │       0 B
      res │ 12.9 KiB │ 12.9 KiB │       0 B │  8.5 KiB │  8.5 KiB │       0 B
    asset │  3.2 KiB │  2.7 KiB │    -543 B │  2.9 KiB │  2.4 KiB │    -539 B
    other │ 41.4 KiB │ 41.4 KiB │       0 B │ 40.6 KiB │ 40.6 KiB │       0 B
──────────┼──────────┼──────────┼───────────┼──────────┼──────────┼───────────
    total │  2.2 MiB │  2.3 MiB │ +13.2 KiB │  2.3 MiB │  2.3 MiB │ +13.2 KiB


 DEX     │ old   │ new   │ diff
─────────┼───────┼───────┼─────────────────
   files │     1 │     1 │   0
 strings │ 17115 │ 17157 │ +42 (+55 -13)
   types │  3671 │  3683 │ +12 (+21 -9)
 classes │  2950 │  2962 │ +12 (+21 -9)
 methods │ 14174 │ 14272 │ +98 (+414 -316)
  fields │  8491 │  8525 │ +34 (+111 -77)


 ARSC    │ old │ new │ diff
─────────┼─────┼─────┼──────
 configs │  19 │  19 │  0
 entries │ 260 │ 260 │  0


=================
====   APK   ====
=================

     compressed      │    uncompressed     │
─────────┬───────────┼─────────┬───────────┤
 size    │ diff      │ size    │ diff      │ path
─────────┼───────────┼─────────┼───────────┼────────────────────────────────
 2.2 MiB │ +13.7 KiB │ 2.2 MiB │ +13.7 KiB │ ∆ classes.dex
 2.3 KiB │    -542 B │ 2.1 KiB │    -542 B │ ∆ assets/dexopt/baseline.prof
   401 B │      -1 B │   267 B │      +3 B │ ∆ assets/dexopt/baseline.profm
─────────┼───────────┼─────────┼───────────┼────────────────────────────────
 2.2 MiB │ +13.2 KiB │ 2.2 MiB │ +13.2 KiB │ (total)

=================
====   DEX   ====
=================

TYPES:

   old  │ new  │ diff
  ──────┼──────┼──────────────
   3671 │ 3683 │ +12 (+21 -9)

  + Landroidx/activity/compose/BackHandlerKt$BackHandler$3;
  + Landroidx/compose/ui/layout/RootMeasurePolicy$measure$4;
  + Landroidx/room/AutoCloser$Companion;
  + Lcom/jakewharton/sa4p/presenter/MainPresenterKt$mainPresenter$1$1;
  + Lcom/jakewharton/sa4p/presenter/MainPresenterKt$mainPresenter$3$1;
  + Lcom/jakewharton/sa4p/presenter/MainPresenterKt$mainPresenter$3;
  + Lcom/jakewharton/sa4p/ui/PendingUrlsKt$PendingUrls$2;
  + Lkotlinx/collections/immutable/ImmutableList$SubList;
  + Lkotlinx/collections/immutable/ImmutableList;
  + Lkotlinx/collections/immutable/PersistentList$Builder;
  + Lkotlinx/collections/immutable/PersistentList;
  + Lkotlinx/collections/immutable/implementations/immutableList/AbstractListIterator;
  + Lkotlinx/collections/immutable/implementations/immutableList/AbstractPersistentList;
  + Lkotlinx/collections/immutable/implementations/immutableList/BufferIterator;
  + Lkotlinx/collections/immutable/implementations/immutableList/PersistentVector;
  + Lkotlinx/collections/immutable/implementations/immutableList/PersistentVectorBuilder;
  + Lkotlinx/collections/immutable/implementations/immutableList/PersistentVectorIterator;
  + Lkotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator;
  + Lkotlinx/collections/immutable/implementations/immutableList/SmallPersistentVector;
  + Lkotlinx/collections/immutable/implementations/immutableList/TrieIterator;
  + Lokio/Utf8;

  - Landroidx/compose/foundation/lazy/LazyListItemProviderImpl$Item$2;
  - Landroidx/compose/runtime/external/kotlinx/collections/immutable/implementations/immutableList/AbstractListIterator;
  - Landroidx/core/view/accessibility/AccessibilityNodeProviderCompat;
  - Landroidx/work/SystemClock;
  - Lcom/jakewharton/sa4p/presenter/MainPresenterKt$MainPresenter$1$1;
  - Lcom/jakewharton/sa4p/presenter/MainPresenterKt$MainPresenter$3$1;
  - Lcom/jakewharton/sa4p/presenter/MainPresenterKt$MainPresenter$3;
  - Lcom/jakewharton/sa4p/ui/PendingUrlsKt$PendingUrls$1;
  - Lokio/ByteString$Companion;
```

File: app/src/main/kotlin/com/jakewharton/sa4p/ui/MainUi.kt [hunk 1]
```diff
@@ -33,7 +37,7 @@ fun MainUi(model: MainModel) {
 			},
 		) {
 			Box(modifier = Modifier.padding(it)) {
-				PendingUrls(model.pendingUrls)
+				PendingUrls(model.pendingUrls.toPersistentList())
```
Comment (JakeWharton) [inline app/src/main/kotlin/com/jakewharton/sa4p/ui/MainUi.kt]: This conversion to an immutable list should be done when creating the model at the presenter layer, not as part of UI rendering.

File: app/src/main/kotlin/com/jakewharton/sa4p/ui/MainUi.kt [hunk 2]
```diff
@@ -13,10 +13,14 @@ import androidx.compose.ui.Modifier
...
 import com.jakewharton.sa4p.R
 import com.jakewharton.sa4p.presenter.MainModel
+import kotlinx.collections.immutable.toPersistentList
 
 @OptIn(ExperimentalMaterial3Api::class)
 @Composable
-fun MainUi(model: MainModel) {
+fun MainUi(
+	model: MainModel,
+	modifier: Modifier = Modifier,
```
Comment (JakeWharton) [inline app/src/main/kotlin/com/jakewharton/sa4p/ui/MainUi.kt:21]: This is unused. Either we need to hook it up correctly to the scaffold or suppress the lint check.

File: app/src/main/kotlin/com/jakewharton/sa4p/ui/footer.kt [hunk 3]
```diff
@@ -35,7 +35,10 @@ import com.jakewharton.sa4p.presenter.MainModel
 import com.jakewharton.sa4p.presenter.Unauthenticated
 
 @Composable
-fun BottomBar(model: MainModel) {
+fun BottomBar(
+	model: MainModel,
+	modifier: Modifier = Modifier,
```
Comment (JakeWharton) [inline app/src/main/kotlin/com/jakewharton/sa4p/ui/footer.kt:40]: Same as above. Needs to be used by the root layout.

File: app/src/main/kotlin/com/jakewharton/sa4p/ui/pendingUrls.kt [hunk 4]
```diff
@@ -6,13 +6,19 @@ import androidx.compose.material3.Divider
...
 import com.jakewharton.sa4p.db.Urls
+import kotlinx.collections.immutable.ImmutableList
+import kotlinx.collections.immutable.persistentListOf
 import kotlinx.datetime.Instant
 
 @Composable
-fun PendingUrls(urls: List<Pending>) {
+fun PendingUrls(
+	urls: ImmutableList<Pending>,
+	modifier: Modifier = Modifier,
```
Comment (JakeWharton) [inline app/src/main/kotlin/com/jakewharton/sa4p/ui/pendingUrls.kt:20]: Needs to be used by the list item and lazy column branches.

File: app/src/main/kotlin/com/jakewharton/sa4p/ui/theme.kt [hunk 5]
```diff
@@ -137,6 +138,7 @@ private val DarkColors = darkColorScheme(
 
 @Composable
 fun Theme(
+	modifier: Modifier = Modifier,
```
Comment (JakeWharton) [inline app/src/main/kotlin/com/jakewharton/sa4p/ui/theme.kt]: Not sure this needs a modifier. Does `MaterialTheme` take one?

**Review by JakeWharton — APPROVED** [review]
Mostly good, just need to hook up the modifiers where added.

File: app/src/main/kotlin/com/jakewharton/sa4p/ui/footer.kt (hunk 3 above)
Comment (Goooler) [reply app/src/main/kotlin/com/jakewharton/sa4p/ui/footer.kt:40]: BottomAppBar has its own modifier.

File: app/src/main/kotlin/com/jakewharton/sa4p/ui/theme.kt (hunk 5 above)
Comment (Goooler) [reply app/src/main/kotlin/com/jakewharton/sa4p/ui/theme.kt]: MaterialTheme doesn't have a modifier.

File: app/src/main/kotlin/com/jakewharton/sa4p/ui/footer.kt [hunk 6]
```diff
@@ -34,8 +34,11 @@ import com.jakewharton.sa4p.presenter.Authenticated
...
 import com.jakewharton.sa4p.presenter.Unauthenticated
 
+@Suppress("ktlint:compose:modifier-missing-check") // BottomAppBar has its own modifier
 @Composable
-fun BottomBar(model: MainModel) {
+fun BottomBar(
+	model: MainModel,
+) {
 	BottomAppBar(
 		modifier = Modifier.wrapContentHeight(),
```
Comment (JakeWharton) [inline app/src/main/kotlin/com/jakewharton/sa4p/ui/footer.kt]: I still think this should take a parameter, but instead you use it as the initial modifier value.
```suggestion
		modifier = modifier.wrapContentHeight(),
```

Comment (Goooler) [reply app/src/main/kotlin/com/jakewharton/sa4p/ui/footer.kt]: I recalled that applying this would conflict with another rule, but it works fine now.

**Pull request merged**